controllers/gce/main.go

```diff
@@ -245,58 +245,107 @@ func main() {
 	// We don't watch because we know exactly which configmap we want and this
 	// controller already watches 5 other resources, so it isn't worth the cost
 	// of another connection and complexity.
 	go wait.Forever(func() {
-		uid, found, err := vault.Get()
-		existing := namer.GetClusterName()
-		if found && uid != existing {
-			glog.Infof("Cluster uid changed from %v -> %v", existing, uid)
-			namer.SetClusterName(uid)
-		} else if err != nil {
-			glog.Errorf("Failed to reconcile cluster uid %v, currently set to %v", err, existing)
+		for _, key := range [...]string{storage.UidDataKey, storage.ProviderDataKey} {
+			val, found, err := uidVault.Get(key)
+			if err != nil {
+				glog.Errorf("Can't read uidConfigMap %v", uidConfigMapName)
+			} else if !found {
+				glog.Errorf("Can't read %v from uidConfigMap %v", key, uidConfigMapName)
+				continue
+			}
+
+			switch key {
+			case storage.UidDataKey:
+				if uid := namer.GetClusterName(); uid != val {
+					glog.Infof("Cluster uid changed from %v -> %v", uid, val)
+					namer.SetClusterName(val)
+				}
+			case storage.ProviderDataKey:
+				if fw_name := namer.GetFirewallName(); fw_name != val {
+					glog.Infof("Cluster firewall name changed from %v -> %v", fw_name, val)
+					namer.SetFirewallName(val)
+				}
+			}
 		}
 	}, 5*time.Second)
 	return namer, nil
 }
 
-// getClusterUID returns the cluster UID. Rules for UID generation:
-// If the user specifies a --cluster-uid param it overwrites everything
-// else, check UID config map for a previously recorded uid
-// else, check if there are any working Ingresses
-//	- remember that "" is the cluster uid
-// else, allocate a new uid
-func getClusterUID(kubeClient client.Interface, name string) (string, error) {
-	cfgVault := storage.NewConfigMapVault(kubeClient, api.NamespaceSystem, uidConfigMapName)
+// getFlagOrLookupVault returns the name to use associated to a flag and configmap.
+// The returned value follows this priority:
+// If the provided 'name' is not empty, that name is used.
+//       This is effectively a client override via a command line flag.
+// else, check configmap under 'configmap_name' as a key and if found, use the associated value
+// else, return an empty 'name' and pass along an error iff the configmap lookup is erroneous.
+func getFlagOrLookupVault(cfgVault *storage.ConfigMapVault, cm_key string, name string) (string, error) {
 	if name != "" {
-		glog.Infof("Using user provided cluster uid %v", name)
+		glog.Infof("Using user provided %v %v", cm_key, name)
 		// Don't save the uid in the vault, so users can rollback through
-		// --cluster-uid=""
+		// setting the accompany flag to ""
 		return name, nil
 	}
-
-	existingUID, found, err := cfgVault.Get()
+	val, found, err := cfgVault.Get(cm_key)
 	if found {
-		glog.Infof("Using saved cluster uid %q", existingUID)
-		return existingUID, nil
+		glog.Infof("Using %v = %q saved in ConfigMap", cm_key, val)
+		return val, nil
 	} else if err != nil {
```

Comment:
This is an master-existing nitpick of mine. Could we reorder the `err != nil` before the `if found`? 

Thank you for fixing it in the `newNamer`! https://github.com/kubernetes/ingress/blob/master/controllers/gce/main.go#L269